File: examples/FSharp/OpenGL/Tutorial 1.3 - Abstractions/BufferObject.fs
#nowarn "9"
namespace Tutorial

open Microsoft.FSharp.NativeInterop
open Silk.NET.OpenGL

type BufferObject = {
    gl: GL
    handle: uint32
    bufferType: GLEnum
}

module BufferObject =
    let bind (bufferObject: BufferObject) =
        bufferObject.gl.BindBuffer(bufferObject.bufferType, bufferObject.handle)


    let create (gl: GL) (data: 'DataType[]) (bufferType: GLEnum) =
        use ptr = fixed data
        let voidPtr = NativePtr.toVoidPtr ptr
        let size = uint32 (data.Length * sizeof<DataType>)
        
        let handle = gl.GenBuffer()
        let bufferObject = {
            gl = gl
            handle = handle
            bufferType = bufferType
        }

        bind bufferObject
        gl.BufferData(bufferType, size, voidPtr, GLEnum.StaticDraw)

        bufferObject

    let destroy (bufferObject: BufferObject) =
        bufferObject.gl.DeleteBuffer(bufferObject.handle)

File: examples/CSharp/OpenGL Tutorials/Tutorial 4.1 - Model Loading/BufferObject.cs
using Silk.NET.OpenGL;
using System;

namespace Tutorial
{
    public class BufferObject<TDataType> : IDisposable
        where TDataType : unmanaged
    {
        private uint _handle;
        private BufferTargetARB _bufferType;
        private GL _gl;

        public unsafe BufferObject(GL gl, Span<TDataType> data, BufferTargetARB bufferType)
        {
            _gl = gl;
            _bufferType = bufferType;

            _handle = _gl.GenBuffer();
            Bind();
            fixed (void* d = data)
            {
                _gl.BufferData(bufferType, (nuint) (data.Length * sizeof(TDataType)), d, BufferUsageARB.StaticDraw);
            }
        }

        public void Bind()
        {
            _gl.BindBuffer(_bufferType, _handle);
        }

        public void Dispose()
        {
            _gl.DeleteBuffer(_handle);
        }
    }
}
